azkaban-common/src/main/java/azkaban/server/HttpRequestUtils.java

```diff
@@ -97,12 +101,20 @@ public static ExecutionOptions parseFlowOptions(final HttpServletRequest req)
       execOptions.setPipelineLevel(queueLevel);
     }
 
-    String mailCreator = DefaultMailCreator.DEFAULT_MAIL_CREATOR;
     if (hasParam(req, "mailCreator")) {
-      mailCreator = getParam(req, "mailCreator");
+      final String mailCreator = getParam(req, "mailCreator");
       execOptions.setMailCreator(mailCreator);
     }
 
+    final Map<String, String> slaSettings = getParamGroup(req, PARAM_SLA_SETTINGS);
+    // emails param is optional
+    final String emailStr = getParam(req, PARAM_SLA_EMAILS, null);
+    final List<SlaOption> slaOptions = SlaRequestUtils.parseSlaOptions(flowName, emailStr,
+        slaSettings);
```

Comment:
_parseSlaOptions_ throws an exception if a SLA rule is specified without any actions (for example: send an email, kill execution). How are those validation errors notified to the user?

Also, I noticed _SlaRequestUtils.parseSlaSetting_ throws a `ScheduleManagerException`. Will `ServletException` (in the absence of validation exceptions) be more appropriate for this utility method?

Reply:
> How are those validation errors notified to the user?

Same as usual:

https://github.com/azkaban/azkaban/blob/2a3f861d33f5df9a251086837a6eeb7e80f75614/azkaban-web-server/src/main/java/azkaban/webapp/servlet/ScheduleServlet.java#L200-L201

> I noticed SlaRequestUtils.parseSlaSetting throws a ScheduleManagerException

Not true. `SlaRequestUtils.parseSlaSetting` throws `ServletException`:

https://github.com/azkaban/azkaban/blob/2a3f861d33f5df9a251086837a6eeb7e80f75614/azkaban-common/src/main/java/azkaban/server/SlaRequestUtils.java#L22-L23

Reply:
Thanks for generalizing the exception thrown in the utility class. 

SLA validation errors are handled correctly when creating schedules as you pointed out but not when creating a flow execution. `ExecutorServlet::ajaxExecuteFlow()` replies with a 500 response if `HttpRequestUtils.parseFlowOptions(req)` throws an exception. Could you make the necessary changes to handle validation errors there as well?

Reply:
Nicely spotted. I'll add it.

Btw, did you have some set of commands for manually testing `ExecutorServlet::ajaxExecuteFlow()` like that?

Reply:
sorry, I don't have any fancy commands :)

Reply:
Ookay, so did you deduce that there would be a 500 response simply by looking at the code?

Reply:
yes, while working on apis changes, I've seen this is what happens when exceptions just bubble up.